moveit_core/collision_detection_bullet/src/collision_robot_bt.cpp

```diff
@@ -0,0 +1,346 @@
+
+#include <moveit/collision_detection_bullet/collision_robot_bt.h>
+#include <moveit/collision_detection_bullet/fcl_compat.h>
+#include <tesseract_ros/ros_tesseract_utils.h>
+
+#if (MOVEIT_FCL_VERSION >= FCL_VERSION_CHECK(0, 6, 0))
+#include <fcl/broadphase/broadphase_dynamic_AABB_tree.h>
+#endif
+
+namespace collision_detection
+{
+CollisionRobotBt::CollisionRobotBt(const robot_model::RobotModelConstPtr& model, double padding, double scale)
+  : CollisionRobot(model, padding, scale)
+{
+  const std::vector<const robot_model::LinkModel*>& links = robot_model_->getLinkModelsWithCollisionGeometry();
+  std::size_t index;
+  geoms_.resize(robot_model_->getLinkGeometryCount());
+  fcl_objs_.resize(robot_model_->getLinkGeometryCount());
+  // we keep the same order of objects as what RobotState *::getLinkState() returns
+  for (auto link : links)
+    for (std::size_t j = 0; j < link->getShapes().size(); ++j)
+    {
+      FCLGeometryConstPtr g = createCollisionGeometry(link->getShapes()[j], getLinkScale(link->getName()),
+                                                      getLinkPadding(link->getName()), link, j);
+      if (g)
+      {
+        index = link->getFirstCollisionBodyTransformIndex() + j;
+        geoms_[index] = g;
+
+        // Need to store the FCL object so the AABB does not get recreated every time.
+        // Every time this object is created, g->computeLocalAABB() is called  which is
+        // very expensive and should only be calculated once. To update the AABB, use the
+        // collObj->setTransform and then call collObj->computeAABB() to transform the AABB.
+        fcl_objs_[index] = FCLCollisionObjectConstPtr(new fcl::CollisionObjectd(g->collision_geometry_));
+      }
+      else
+        ROS_ERROR_NAMED("collision_detection.bullet", "Unable to construct collision geometry for link '%s'",
+                        link->getName().c_str());
+    }
+
+  // this is from tesseract when creating a new KDLenv
+  // ---------------
+  for (const auto& link : robot_model_->getURDF()->links_)
+  {
+    if (link.second->collision_array.size() > 0)
+    {
+      const std::vector<urdf::CollisionSharedPtr>& col_array =
+          link.second->collision_array.empty() ? std::vector<urdf::CollisionSharedPtr>(1, link.second->collision) :
+                                                 link.second->collision_array;
+
+      std::vector<shapes::ShapeConstPtr> shapes;
+      tesseract::VectorIsometry3d shape_poses;
+      tesseract::CollisionObjectTypeVector collision_object_types;
+
+      for (std::size_t i = 0; i < col_array.size(); ++i)
+      {
+        if (col_array[i] && col_array[i]->geometry)
+        {
+          shapes::ShapeConstPtr s = tesseract::tesseract_ros::constructShape(col_array[i]->geometry.get());
+          if (s)
+          {
+            shapes.push_back(s);
+            shape_poses.push_back(tesseract::tesseract_ros::urdfPose2Eigen(col_array[i]->origin));
+
+            if (s->type == shapes::MESH)
```

Comment:
Add a `TODO(j-petit)` here, as eventually we want to distinguish between convex and non-convex meshes (or at least read if the mesh is convex and spit out a warning if it's not).